contracts/carrot-app/src/handlers/execute.rs

```diff
@@ -346,53 +346,44 @@ pub fn autocompound_executor_rewards(
     let user = position.owner;
 
     // Get user balance of gas denom
-    let user_gas_balance = deps
-        .querier
-        .query_balance(user.clone(), rewards_config.gas_denom.clone())?;
+    let gas_denom = rewards_config
+        .gas_asset
+        .resolve(&deps.querier, &app.ans_host(deps)?)?;
+    let user_gas_balance = gas_denom.query_balance(&deps.querier, user.clone())?;
 
     let mut rewards_messages = vec![];
 
     // If not enough gas coins - swap for some amount
-    if user_gas_balance.amount < rewards_config.min_gas_balance {
+    if user_gas_balance < rewards_config.min_gas_balance {
         // Get asset entries
         let dex = app.ans_dex(deps, OSMOSIS.to_string());
-        let ans_host = app.ans_host(deps)?;
-        let gas_asset = AssetInfo::Native(rewards_config.gas_denom.clone())
-            .resolve(&deps.querier, &ans_host)?;
-        let swap_asset = AssetInfo::Native(rewards_config.swap_denom.clone())
-            .resolve(&deps.querier, &ans_host)?;
 
         // Do reverse swap to find approximate amount we need to swap
-        let need_gas_coins = rewards_config.max_gas_balance - user_gas_balance.amount;
+        let need_gas_coins = rewards_config.max_gas_balance - user_gas_balance;
         let simulate_swap_response = dex.simulate_swap(
-            AnsAsset::new(gas_asset.clone(), need_gas_coins),
-            swap_asset.clone(),
+            AnsAsset::new(rewards_config.gas_asset.clone(), need_gas_coins),
+            rewards_config.swap_asset.clone(),
         )?;
 
         // Get user balance of swap denom
-        let user_swap_balance = deps
-            .querier
-            .query_balance(user.clone(), rewards_config.swap_denom)?;
+        let user_swap_balance =
+            get_balance(rewards_config.swap_asset.clone(), deps, user.clone(), app)?;
 
         // Swap as much as available if not enough for max_gas_balance
-        let swap_amount = simulate_swap_response
-            .return_amount
-            .min(user_swap_balance.amount);
+        let swap_amount = simulate_swap_response.return_amount.min(user_swap_balance);
 
         let msgs = swap_msg(
             deps,
             env,
-            AnsAsset::new(swap_asset, swap_amount),
-            gas_asset,
+            AnsAsset::new(rewards_config.swap_asset, swap_amount),
+            rewards_config.gas_asset,
             app,
         )?;
         rewards_messages.extend(msgs);
     }
 
-    let reward = Coin {
-        denom: rewards_config.gas_denom,
-        amount: rewards_config.reward,
-    };
+    let reward: Coin = Asset::new(gas_denom, rewards_config.reward).try_into()?;
+
```

Comment:
Ok, we will support non-native rewards in this function for now. @CyberHoward, do you think we should error on validation in instantiatemsg ?

Reply:
Yeah I think that's fine. I don't think we need to over-optimize the bot-reward logic. Croncat built a whole product to do this. If we ever need more control we should use that instead of re-implementing our own version.